File: sdk/typescript/src/delegation.ts
/**
 * Automatic delegation chain tracking via AsyncLocalStorage.
 * Propagates through async/await, Promises, and callbacks automatically.
 *
 * When Agent A's guarded tool calls Agent B's guarded tool in the same
 * process, the SDK auto-detects the cross-agent hop and includes
 * delegation metadata in proxy requests.
 */
import { AsyncLocalStorage } from "node:async_hooks";
import { randomUUID } from "node:crypto";

export interface DelegationContext {
  traceId: string;
  chain: string[];
  confidence: "verified" | "inferred" | "declared";
}

const delegationStore = new AsyncLocalStorage<DelegationContext>();

export const MAX_DELEGATION_DEPTH = 5;

/**
 * Get the current delegation context, if any.
 */
export function getDelegation(): DelegationContext | undefined {
  return delegationStore.getStore();
}

/**
 * Get the caller (parent) agent ID from the delegation chain.
 * Returns undefined if there is no parent (single agent or no context).
 */
export function getCallerAgentId(): string | undefined {
  const ctx = delegationStore.getStore();
  if (!ctx || ctx.chain.length < 2) return undefined;
  return ctx.chain[ctx.chain.length - 2];
}

/**
 * Check whether a chain contains a cycle (duplicate agent ID).
 */
export function hasCycle(chain: string[]): boolean {
  return new Set(chain).size !== chain.length;
}

/**
 * Run a function within a delegation context.
 * If already inside a delegation, extends the chain.
 * If not, starts a new chain.
 */
export function withDelegation<T>(agentId: string, fn: () => T): T {
  const parent = delegationStore.getStore();

  let ctx: DelegationContext;
  if (parent) {
    ctx = {
      traceId: parent.traceId,
      chain: [...parent.chain, agentId],
      confidence: parent.confidence,
    };
  } else {
    ctx = {
      traceId: randomUUID().replace(/-/g, "").slice(0, 16),
      chain: [agentId],
      confidence: "verified",
    };
  }

  if (ctx.chain.length > MAX_DELEGATION_DEPTH) {
    throw new Error(
      `Delegation depth ${ctx.chain.length} exceeds maximum ${MAX_DELEGATION_DEPTH}. ` +
        `Chain: ${ctx.chain.join(" \u2192 ")}`,
    );
  }

  if (hasCycle(ctx.chain)) {
    throw new Error(
      `Circular delegation detected: ${ctx.chain.join(" \u2192 ")}`,
    );
  }

  return delegationStore.run(ctx, fn);
}

/**
 * Get delegation headers for cross-service HTTP propagation.
 * Returns an empty object when called outside a delegation context.
 */
export function delegationHeaders(): Record<string, string> {
  const ctx = delegationStore.getStore();
  if (!ctx || ctx.chain.length < 2) return {};
  return {
    "X-Clampd-Delegation-Trace": ctx.traceId,
    "X-Clampd-Delegation-Chain": ctx.chain.join(","),
    "X-Clampd-Delegation-Confidence": ctx.confidence,
  };
}

File: sdk/python/clampd/delegation.py
"""Automatic delegation chain tracking via contextvars."""
from __future__ import annotations

import contextvars
import uuid
from dataclasses import dataclass, field

MAX_DELEGATION_DEPTH = 5


@dataclass
class DelegationContext:
    """Tracks the delegation chain for cross-agent calls."""
    trace_id: str = field(default_factory=lambda: uuid.uuid4().hex[:16])
    chain: list[str] = field(default_factory=list)
    confidence: str = "verified"  # verified | inferred | declared

    @property
    def depth(self) -> int:
        return len(self.chain)

    @property
    def caller_agent_id(self) -> str | None:
        """The agent that initiated this delegation hop."""
        return self.chain[-2] if len(self.chain) >= 2 else None

    def has_cycle(self) -> bool:
        return len(self.chain) != len(set(self.chain))


# Per-request delegation context — propagates through async/await automatically
_delegation_ctx: contextvars.ContextVar[DelegationContext | None] = contextvars.ContextVar(
    'clampd_delegation', default=None
)


def get_delegation() -> DelegationContext | None:
    """Get current delegation context (if inside a delegated call)."""
    return _delegation_ctx.get()


def enter_delegation(agent_id: str) -> tuple[DelegationContext, contextvars.Token[DelegationContext | None]]:
    """Enter a delegation scope. Returns context and reset token."""
    parent = _delegation_ctx.get()
    if parent is not None:
        # Nested call — extend the chain
        ctx = DelegationContext(
            trace_id=parent.trace_id,
            chain=parent.chain + [agent_id],
            confidence=parent.confidence,
        )
    else:
        # Root call — start a new chain
        ctx = DelegationContext(
            trace_id=uuid.uuid4().hex[:16],
            chain=[agent_id],
        )
    token = _delegation_ctx.set(ctx)
    return ctx, token


def exit_delegation(token: contextvars.Token[DelegationContext | None]) -> None:
    """Exit delegation scope, restoring previous context."""
    _delegation_ctx.reset(token)
